### PR TITLE
Updating mnemonic language determination

### LINKED ISSUE #119: Common characters being overridden in chinese simplified and chinese Traditional word lists
From https://github.com/ethereum/staking-deposit-cli/issues/410

“While going through the code, this code [here](https://github.com/ethereum/staking-deposit-cli/blob/master/staking_deposit/key_handling/key_derivation/mnemonic.py#L70) is causing a problem but hiding it. There are common characters between the 2 word lists and when the map is created in the referenced code, if there is a collision in keys, it is silently overridden to the new file being read. This issue can be seen if the code checks for collision of keys.”

We need to write a test to catch this, and fix the issue.

### PR DESCRIPTION
**Pull request by valefar-on-discord**
Fixes Closes Resolves #119 

## Changes

Updating the language determination of mnemonics to allow a list of languages per word

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

Creating mnemonics with Chinese Simplified or Chinese Traditional can easily reproduce the issue with a set_trace in the `determine_mnemonic_language` logic

#### Requires documentation update

- [ ] Yes
- [X] No

#### Requires explanation in Release Notes

- [ ] Yes
- [X] No

## Remarks

The issue here is the word language map only allows for a single language per word. Because there is so much overlap between Chinese Simplified and Chinese Traditional, you can run into a situation where, after generating a mnemonic in Chinese Simplified, your word map could look like this:

```
ipdb> word_languages
[['chinese_traditional'], ['chinese_traditional'], ['chinese_simplified'], ['chinese_traditional'], ['chinese_traditional'], 
['chinese_traditional'], ['chinese_traditional'], ['chinese_traditional'], ['chinese_traditional'], ['chinese_simplified'], 
['chinese_simplified'], ['chinese_traditional'], ['chinese_simplified'], ['chinese_traditional'], ['chinese_simplified'], 
['chinese_traditional'], ['chinese_simplified'], ['chinese_traditional'], ['chinese_traditional'], ['chinese_traditional'], 
['chinese_traditional'], ['chinese_traditional'], ['chinese_traditional'], ['chinese_simplified']]
```

But even in this situation, the logic will go through both discovered languages and provide the correct mnemonic. 

The failure case would be if you generate a mnemonic with Chinese Simplified, but each word is also in Chinese Traditional and results in an invalid mnemonic.

If that scenario is even possible in the bounds of how mnemonics work then it would have to be from a mnemonic created outside this tool as there is a verification step in the `new-mnemonic` flow. 

Either way, the logic has been updated to allow for multiple languages per word resulting in this for the same test mnemonic:
```
ipdb> word_languages
[['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified'], 
['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 
'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], 
['chinese_simplified', 'chinese_traditional'], ['chinese_simplified'], ['chinese_simplified'], ['chinese_simplified', 
'chinese_traditional'], ['chinese_simplified'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified'], 
['chinese_simplified', 'chinese_traditional'], ['chinese_simplified'], ['chinese_simplified', 'chinese_traditional'], 
['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 
'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], ['chinese_simplified', 'chinese_traditional'], 
['chinese_simplified']]
```

### CODE REVIEW

**Comment by valefar-on-discord** [comment]
Every test failed in the same manner: `This mnemonic abbreviated form is available in multiple languages`

This happens because some mnemonics, `的 的 的 的 的 的 的 的 的 的 的 在` as an example, are the same in both Chinese Simplified and Chinese Traditional.

Prior to this change the resulting word languages would have only contained Chinese Traditional as all instances of Chinese Simplified would have been overwritten.  

With this change both are now returned and both result in a valid mnemonic which then raises the error.

I'm not sure how we want to handle this.  Though this is a little spooky, I don't believe it is possible for the underlying private key to differ if there is duplication between languages but having that confirmed by someone more knowledgeable than I would be preferred.

**Comment by yorickdowne** [comment]
> The failure case would be if you generate a mnemonic with Chinese Simplified, but each word is also in Chinese Traditional and results in an invalid mnemonic.

Do we know that this failure case is possible? I want to be sure we are solving something that can actually happen. Did the original report give us a sample mnemonic that causes a failure, that is, two different keys in Simplified and Traditional?

**Comment by remyroy** [comment]
I'm trying to wrap my head around this.

In the `new-mnemonic` flow, we try to reconstruct the mnemonic from words reentered after the user already selected a language. I think we should restrict the reentered mnemonic language to the one already selected instead of trying to *guess* it again. So in https://github.com/eth-educators/ethstaker-deposit-cli/blob/main/ethstaker_deposit/cli/new_mnemonic.py#L52, we could pass the `mnemonic_language` down the `reconstruct_mnemonic` function for this. This would still support entering the abreviated form of the mnemonic when confirming it in that flow.

This would leave the `existing-mnemonic` flow remaining for this issue where we try to *guess* the mnemonic language in https://github.com/eth-educators/ethstaker-deposit-cli/blob/main/ethstaker_deposit/cli/existing_mnemonic.py#L60 . In that case, we could have a potentially conflict where the words entered match multiple languages. I think we could more elegantly solve this in one or multiple ways:

1. In case we are not sure, we could prompt the user with something like *The mnemonic you entered matches multiple langauges, which ones would you like to use for this command? [1. 简体中文, 2. 繁體中文,]:  [简体中文]:*
2. We could add another flag `--mnemonic_language` to the `existing-mnemonic` command to force using that language.
3. We could simply fail with that message above `This mnemonic abbreviated form is available in multiple languages`. This has the con that if someone really has one of those potentially conflicting mnemonics that match to multiple languages that he or she wouldn't be able to use it in the `existing-mnemonic` flow. This would seems like a bad outcome to me.

What do you think @valefar-on-discord ? Is this a good picture of the whole situation? Am I missing something here?

**Comment by valefar-on-discord** [comment]
I talked to a cryptographic expert and they let me know that the mnemonic word is used as a lookup to determine the index in the word list.  As long as the matching words have the identical index in the corresponding word list, the underlying private key and cryptographic results will be identical.

I did a comparison of Chinese simplified/traditional and, though there are 1275 shared words, each one are of the same index.

My guess is this is a non-issue and the creation of these word lists has this consideration in mind.

That's not to say we should be complicit. I am a fan of Remy's prompt solution and it's not that big of a lift.

**Comment by yorickdowne** [comment]
Does prompting improve UX? It may be simpler to just make an assumption either way, if the resulting keys are the same anyway.

**Comment by yorickdowne** [comment]
Here’s what we have from the reporting user. Let’s see how this comes out in simplified and traditional Chinese.

> This was one of the vectors that had failed.
的 的 的 的 的 的 的 的 的 的 的 在

**Comment by chong-he** [comment]
Is the user trying to recover the private key with another tool other than the deposit-cli? Because with `./deposit.sh existing-mnemonic`, there is no option of Chinese Traditional (CHT), only Chinese Simplified (CHS): 12. 简 体中文

For mnemonic like: 的 的 的 的 的 的 的 的 的 的 的 在 where the mnemonic is valid for both languages CHS and CHT, the resulting private key should be the same. I remember it is something like: 

12 words --> 12 indices (12 numbers from 0-2047) --> using these 12 numbers to generate the keys  

As @valefar-on-discord mentioned, each word corresponds to an index in the word list. I have also checked and can confirm that when CHS and CHT are the same word, the indices are the same. Therefore, the resulting key should be the same, because they essentially have the same 12 indices/numbers. If there is some tools to verify this, then we can make an assumption either way as @yorickdowne says.

Curious what tools did @anupsv used that encounter this issue?

**Comment by chong-he** [comment]
> @chong-he The prompt you are seeing is for the language used for the tool's prompts and outputs. This allows for 12 languages that are partially translated.
> 
> When generating a mnemonic, you are given the mnemonic language option of `Please choose the language of the mnemonic word list [1. 简体中文, 2. 繁體中文, 3. čeština, 4. English, 5. Français, 6. Italiano, 7. 日本語, 8. 한국어, 9. Português, 10. Español]: [english]:` Where option 1 is Chinese Simplified and 2 is Chinese Traditional.
> 
> @yorickdowne In your example, because multiple languages are determined for the list of words, each language will be attempted and ultimately output:
> 
> * A `ValidatorError` complaining the abbreviated form exists in multiple languages
> * The fully constructed mnemonic
> * `None` signifying an invalid mnemonic
> 
> Based on how this logic currently exists, Portuguese will not overwrite Spanish but vice-versa, so lets say you generated a Portuguese mnemonic and a single word existed in Spanish: Your word map would be 23 entries containing Portuguese and 1 of Spanish. Two languages will be determined and both iterated upon. Portuguese would result in a valid mnemonic but Spanish would not and the Portuguese full word mnemonic would be returned.
> 
> I ran a script to check the indices of each word across each language to see if there are any that contain mis-matched indices and there are specifically between English and French:
> 
> languageA	languageB	wordA	wordB	indexA	indexB
> english	french	abandon	abandon	0	1
> english	french	amateur	amateur	61	88
> english	french	angle	angle	70	107
> english	french	animal	animal	72	110
> english	french	aspect	aspect	107	148
> english	french	badge	badge	139	190
> english	french	bicycle	bicycle	175	230
> english	french	bonus	bonus	203	262
> english	french	brave	brave	218	277
> english	french	canal	canal	264	323
> english	french	capable	capable	271	328
> english	french	caution	caution	293	347
> english	french	civil	civil	332	403
> english	french	client	client	342	409
> english	french	concert	concert	373	436
> english	french	correct	correct	388	451
> english	french	coyote	coyote	397	461
> english	french	crucial	crucial	419	478
> english	french	cruel	cruel	420	479
> english	french	cycle	cycle	438	493
> english	french	danger	danger	443	498
> english	french	digital	digital	495	562
> english	french	distance	distance	509	573
> english	french	double	double	525	594
> english	french	dragon	dragon	528	598
> english	french	effort	effort	565	631
> english	french	essence	essence	618	725
> english	french	exact	exact	626	757
> english	french	excuse	excuse	632	763
> english	french	fatal	fatal	668	795
> english	french	fatigue	fatigue	670	796
> english	french	festival	festival	681	812
> english	french	figure	figure	688	820
> english	french	fortune	fortune	732	854
> english	french	fragile	fragile	739	861
> english	french	fruit	fruit	750	880
> english	french	globe	globe	794	919
> english	french	guide	guide	828	953
> english	french	humble	humble	887	998
> english	french	image	image	906	1011
> english	french	immense	immense	908	1014
> english	french	impact	impact	910	1017
> english	french	innocent	innocent	932	1043
> english	french	intact	intact	940	1053
> english	french	jaguar	jaguar	953	1070
> english	french	junior	junior	969	1093
> english	french	label	label	993	1102
> english	french	lecture	lecture	1016	1123
> english	french	loyal	loyal	1060	1165
> english	french	machine	machine	1068	1178
> english	french	million	million	1125	1248
> english	french	minute	minute	1130	1254
> english	french	miracle	miracle	1131	1255
> english	french	mobile	mobile	1139	1259
> english	french	muscle	muscle	1164	1286
> english	french	nation	nation	1178	1301
> english	french	nature	nature	1179	1302
> english	french	noble	noble	1197	1322
> english	french	notable	notable	1204	1331
> english	french	opinion	opinion	1243	1381
> english	french	orange	orange	1246	1387
> english	french	ozone	ozone	1268	1409
> english	french	palace	palace	1273	1411
> english	french	panda	panda	1275	1416
> english	french	phrase	phrase	1310	1476
> english	french	piano	piano	1312	1478
> english	french	pizza	pizza	1325	1492
> english	french	position	position	1347	1524
> english	french	prison	prison	1368	1548
> english	french	public	public	1384	1567
> english	french	puzzle	puzzle	1399	1576
> english	french	question	question	1404	1580
> english	french	relief	relief	1450	1626
> english	french	rival	rival	1493	1671
> english	french	romance	romance	1500	1674
> english	french	salon	salon	1524	1707
> english	french	science	science	1543	1727
> english	french	sentence	sentence	1567	1748
> english	french	service	service	1569	1756
> english	french	simple	simple	1608	1769
> english	french	social	social	1647	1777
> english	french	source	source	1664	1801
> english	french	spatial	spatial	1668	1805
> english	french	stable	stable	1694	1809
> english	french	surface	surface	1745	1830
> english	french	surprise	surprise	1747	1833
> english	french	suspect	suspect	1750	1836
> english	french	talent	talent	1771	1847
> english	french	train	train	1848	1911
> english	french	tunnel	tunnel	1876	1933
> english	french	unique	unique	1900	1948
> english	french	usage	usage	1916	1954
> english	french	vague	vague	1925	1963
> english	french	valve	valve	1928	1970
> english	french	village	village	1952	2008
> english	french	virus	virus	1956	2014
> english	french	vital	vital	1960	2020
> english	french	volume	volume	1966	2034
> english	french	voyage	voyage	1968	2039
> english	french	wagon	wagon	1970	2041
> Given the amount of duplicates I think this is an issue we want to account for.

Thanks for the explanation! Now I get that the `existing-mnemonic` language option is just to display the language in the UI, but the mnemonic entered can be of any languages.

As from the above, it seems now that CHS and CHT is not an issue already, right? Rather, it is now English or French issue. And Yorick's proposal sounds great - only prompt when detecting valid mnemonic

**Comment by anupsv** [comment]
> Is the user trying to recover the private key with another tool other than the deposit-cli? Because with `./deposit.sh existing-mnemonic`, there is no option of Chinese Traditional (CHT), only Chinese Simplified (CHS): 12. 简 体中文
> 
> For mnemonic like: 的 的 的 的 的 的 的 的 的 的 的 在 where the mnemonic is valid for both languages CHS and CHT (edit: I assume it is valid on CHT too, haven't tested this as deposit-cli can't recover CHT), the resulting private key should be the same. I remember it is something like:
> 
> 12 words --> 12 indices (12 numbers from 0-2047) --> using these 12 numbers to generate the keys
> 
> As @valefar-on-discord mentioned, each word corresponds to an index in the word list. I have also checked and can confirm that when CHS and CHT are the same word, the indices are the same. Therefore, the resulting key should be the same, because they essentially have the same 12 indices/numbers. If there is some tools to verify this, then we can make an assumption either way as @yorickdowne says.
> 
> Curious what tools did @anupsv used that encounter this issue?

To expand this a bit more further on what I tried, the for loop implementations could be more optimized and I had implemented that. But the test vectors started to fail and that's when i noticed the issue. WIth the current way of how the dict mapping is generated between the word and language, there are many cases where keys are overwritten when collisions happened which I encountered first with CHT and CHS primarily because that is the first test case in the mnemonic test vectors.

**Comment by valefar-on-discord** [comment]
> To expand this a bit more further on what I tried, the for loop implementations could be more optimized and I had implemented that. But the test vectors started to fail and that's when i noticed the issue. WIth the current way of how the dict mapping is generated between the word and language, there are many cases where keys are overwritten when collisions happened which I encountered first with CHT and CHS primarily because that is the first test case in the mnemonic test vectors.

Yeah that's exactly what I am seeing as well and the problem is substantially more concerning with English/French. I think we have an agreed path forward on how to resolve this though it is going to be a _bit_ of a lift.

Great catch and thanks for reporting it!

**Comment by valefar-on-discord** [comment]
Added the ability to prompt the user to select a language if multiple are valid for the same mnemonic.

I tried to manipulate the existing mnemonic logic as little as I could but this is _a_ solution.  I'm not 100% thrilled with it so would love some feedback on how this can be improved.

If we can get to a place where we are happy with the solution I'll wrap it in some unit tests.

But if you run `python -m ethstaker_deposit existing-mnemonic` with `的 的 的 的 的 的 的 的 的 的 的 在` you can see the flow.

I haven't done a thorough round of testing so might be some edge cases

**Comment by yorickdowne** [comment]
Everything removed in `tests/test_key_handling/test_key_derivation/test_vectors/mnemonic.json` has a conflict, I take it. When we test this multi-language, we should also test that we end up with the correct private key

*A* solution is good enough for me, tbh. I am not sure we are getting to "a very elegant / beautiful" solution - it's a somewhat messy problem by its nature.

**Comment by yorickdowne** [comment]
@anupsv and @chong-he , if you have time to test against this solution here, that'd be amazing. What should happen is that it prompts you for the source language of the mnemonic, for a mnemonic that creates a valid key in more than one language

**Comment by yorickdowne** [comment]
@valefar-on-discord , we should probably honor `--mnemonic_language`, and not prompt in that case ... as well as not try to detect and use what we are given, even if that means the mnemonic is then detected as invalid.

**Comment by valefar-on-discord** [comment]
> @valefar-on-discord , we should probably honor `--mnemonic_language`, and not prompt in that case ... as well as not try to detect and use what we are given, even if that means the mnemonic is then invalid, in case a user provided the wrong language.

`--mnemonic_language` currently isn't an argument in those situations. The only time that is an option is when a new mnemonic is created.  Which reminds me this is still an issue when the user verifies the mnemonic in that flow 😭

**Comment by yorickdowne** [comment]
I am thinking about --non_interactive scripting, where a prompt is undesirable. We may want the --mnemonic_language here as an optional parameter, because we can’t always reliably detect the language.

**Comment by chong-he** [comment]
Did some testing about CHS and CHT mnemonic:

1. For mnemonic where CHS and CHT characters are exactly the same, it prompts the user to select the language (1 for CHS and 2 for CHT) and then enter the mnemonic.
The public keys generated for both mnemonic are the same
Example: 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 性

2. For mnemonic where CHS and CHT are different (but the same index in the word list), it wouldn't prompt the user to select the language since they are not the same characters.
The public keys generated are different. I must be missing something here, I thought the public keys generated in this case should also be the same?

    Anyway, I think most of the time a user that is using CHS will always use CHS characters, and likewise for CHT (i.e., people wouldn't convert characters from CHS to CHT just like that, it's not something that people usually do)
So they will end up getting the same public key so long as they keep the characters right

For case 1, I think the language prompt is fine solution, it will only prompts when the mnemonic is valid for >1 language (which I think is not often the case)
Using prompt (assuming the detection works well) is easier imo, so that users don't have add an additional flag when entering the mnemonic (simpler and better UX)
Under what case can we not reliably detect the language?

**Comment by yorickdowne** [comment]
Thanks for helping us test, @chong-he !

We cannot reliably detect the language when:
- Each word in the mnemonic is valid in two or more languages
- The checksum (last word) is valid in two or more languages

For CHS and CHT the indices are the same, so it might be the same key regardless and arguably we don’t need to prompt. But then, why did we get a report - I am cautious whether it’s really the same key.

For English and French the indices are not the same. This should mean that again only one version is correct, as the other one will fail checksum. But, do we know that for certain?

Can I ask, in your testing: You are saying the pub key for CHS is different from the one in CHT, for the same mnemonic? If so, that’s a clear conflict where we need to prompt, exactly what the person originally reporting this was concerned about.

**Comment by valefar-on-discord** [comment]
> I am thinking about --non_interactive scripting, where a prompt is undesirable. We may want the --mnemonic_language here as an optional parameter, because we can’t always reliably detect the language.

I never think of `--non_interactive` 😞 

If we want to add the `--mnemonic_language` argument for that use case then I think a cleaner (code wise) experience would be to exit execution if multiple languages are detected and request the user to provide that argument instead of prompting.

**Comment by yorickdowne** [comment]
That might be cleaner code-wise but isn’t it better UX to prompt if we’re interactive? Overall the promise is that new-mnemonic and existing-mnemonic will work without any additional arguments and prompt for everything they need.

**Comment by valefar-on-discord** [comment]
> That might be cleaner code-wise but isn’t it better UX to prompt if we’re interactive? Overall the promise is that new-mnemonic and existing-mnemonic will work without any additional arguments and prompt for everything they need.

I agree it is a better UX but I question the benefit with the additional code complexity for a use case that, as far as I can tell, is incredibly unlikely to trigger.  I am still not sure if there is such a mnemonic with english and french that would result in different cryptographic values as the underlying mathematics behind it are beyond me.

I created an alternative version that is very similar to this PR but doesn't prioritize UX: https://github.com/eth-educators/ethstaker-deposit-cli/commit/a35d5fd4e0269ae54e6c56fcd5a0928544b5f82a

But if we do want to maximize UX then I can easily add the `--mnemonic_language` as an optional argument doing something similar to: https://github.com/valefar-on-discord/ethstaker-deposit-cli/commit/23a65b6ff0a725fd0fc9c5f4617c1190e110f16f

**Comment by chong-he** [comment]
> Can I ask, in your testing: You are saying the pub key for CHS is different from the one in CHT, for the same mnemonic? If so, that’s a clear conflict where we need to prompt, exactly what the person originally reporting this was concerned about.

Example:
CHS: `缝 塘 诗 浆 遇 困 速 凉 介 西 搭 奖 友 传 吸 释 丧 歌 针 疫 迎 毕 击 盘`
CHT: `縫 塘 詩 漿 遇 困 速 涼 介 西 搭 獎 友 傳 吸 釋 喪 歌 針 疫 迎 畢 擊 盤`

Each word is the same, except that CHS that have their corresponding CHT characters are changed to the CHT characters 
(Same indices word-to-word according to this BIP39 word list: https://github.com/bitcoin/bips/tree/master/bip-0039)

Both mnemonic are valid, but they should be considered different? Because the resulting public keys are different.
I also look up for their word indices, grab their corresponding English language words to form a mnemonic. The resulting mnemonic is valid too, but with different public keys.
This means that even though the 24 word indices are the same, the keys are different for different languages


The original report was, there is a bug with CHS and CHT where when CHS and CHT are having the same characters, the CHS one is being overwritten

So I am thinking, it should be the case that, when a mnemonic that have the exact same words and valid for 2 or more languages, such as:
`的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 性`

Selecting CHS or CHT during key generation/recovery should result in different keys.  I think that's what is meant by the collision mentioned in the original report.

Because if the resulting keys are the same, there is no need to prompt/to have an additional flag (except for the possibility that the keys are different (e.g., English and French), even though the probability of this happening can be very low). 

For now, the public keys generated by ``的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 性` when selecting CHS and CHT during key recovery are the same.

Should they be different in the first place? Because they are the same

File: ethstaker_deposit/exceptions.py [hunk 1]
```diff
@@ -1,2 +1,9 @@
 class ValidationError(Exception):
     ...
+
+
+class MultiLanguageError(Exception):
```
Comment (remyroy) [inline ethstaker_deposit/exceptions.py:5]: Adding this new exception type will need to be handled in https://github.com/eth-educators/ethstaker-deposit-cli/blob/main/ethstaker_deposit/deposit.py#L128 if it bubbles up.

File: ethstaker_deposit/key_handling/key_derivation/mnemonic.py [hunk 2]
```diff
@@ -123,17 +136,17 @@ def reconstruct_mnemonic(mnemonic: str, words_path: str) -> Optional[str]:
...
-                    raise ValidationError("This mnemonic abbreviated form is available in multiple languages.")
+                valid_languages.append(language)
                 reconstructed_mnemonic = ' '.join([_index_to_word(full_word_list, index) for index in word_indices])
             else:
                 pass
         except ValueError:
             pass
+
+    if len(valid_languages) > 1:
+        valid_languages.sort()
```
Comment (remyroy) [inline ethstaker_deposit/key_handling/key_derivation/mnemonic.py:147]: Sorting here should be based not on the languages string value but on the canonical sort order we are using elsewhere which is defined in `ethstaker_deposit/utils/constants.py:MNEMONIC_LANG_OPTIONS`

File: ethstaker_deposit/intl/en/cli/existing_mnemonic.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
     "validate_mnemonic": {
-        "err_invalid_mnemonic": "That is not a valid mnemonic, please check for typos."
+        "err_invalid_mnemonic": "That is not a valid mnemonic, please check for typos.",
+        "arg_mnemonic_language": "Please select from the following languages"
```
Comment (remyroy) [inline ethstaker_deposit/intl/en/cli/existing_mnemonic.json:4]: I think we should expand on this message here. During a normal interactive session of calling something like `uv run python -m ethstaker_deposit existing-mnemonic --mnemonic '的 的 的 的 的 的 的 的 的 的 的 在'`, I get the following output:

```
Please choose your language [1. ‫العربية‬‪ ‬, 2. ελληνικά, 3. English, 4. Français, 5. Bahasa melayu, 6. Italiano, 7. 日本語, 8. 한국어, 9. Português do Brasil, 10. român, 11. Türkçe, 12. 简体中文]:  [English]:

*** Internet connectivity detected ***

To mitigate the risk of unauthorized access and safeguard generated key material, it is strongly advised to run this tool in an offline, airgapped environment. Operating online increases susceptibility to potential theft or compromise.

By continuing, you are accepting responsibility for the risk.

Press any key to continue...
Please select from the following languages [1. 简体中文, 2. 繁體中文]: 1
```

By the time I reach the prompt to select the language for my mnemonic, it is unclear why I need to select a language there or what's wrong compared to a *normal* session.

A better message would be something along the lines of: *We could not determine your mnemonic language. Please select the language you want to use for this mnemonic*. I'm not sure of the exact phrase expression that should be used here, but it should be more clear why we are asking for a language.

**Review by remyroy — CHANGES_REQUESTED** [review]
I would do some minor changes before merging this. Let me know what you think.

File: ethstaker_deposit/key_handling/key_derivation/mnemonic.py (hunk 2 above)
Comment (valefar-on-discord) [reply ethstaker_deposit/key_handling/key_derivation/mnemonic.py:147]: ack

File: ethstaker_deposit/exceptions.py (hunk 1 above)
Comment (valefar-on-discord) [reply ethstaker_deposit/exceptions.py:5]: Great catch

File: ethstaker_deposit/intl/en/cli/existing_mnemonic.json (hunk 3 above)
Comment (valefar-on-discord) [reply ethstaker_deposit/intl/en/cli/existing_mnemonic.json]: Another great point. I have updated the wording to: `We couldn't detect the language of your mnemonic phrase. Please select one of the available language options`

**Review by remyroy — APPROVED** [review]
Everything is fine. I'm merging.
